# **演習：ループとリスト内包表記**

**This notebook is an exercise in the [Python](https://www.kaggle.com/learn/python) course.  You can reference the tutorial at [this link](https://www.kaggle.com/colinmorris/loops-and-list-comprehensions).**

---


**このノートは、[Python] (https://www.kaggle.com/learn/python) コースの練習問題です。チュートリアルは、[このリンク](https://www.kaggle.com/colinmorris/loops-and-list-comprehensions)**で参照できます。

---


With all you've learned, you can start writing much more interesting programs. See if you can solve the problems below.

As always, run the setup code below before working on the questions.

これまで学んできたすべてのことを使って、もっと面白いプログラムを書くことができます。次の問題を解決できるかどうかを確認します。

いつものように、質問に取り組む前に以下のセットアップコードを実行してください。

In [1]:
from learntools.core import binder; binder.bind(globals())
from learntools.python.ex5 import *
print('Setup complete.')

Setup complete.


# 1.

Have you ever felt debugging involved a bit of luck? The following program has a bug. Try to identify the bug and fix it.

# 1.

デバッグにはちょっとした運が必要だと感じたことはありませんか?以下のプログラムにはバグがあります。バグを特定して修正してください。

In [2]:
def has_lucky_number(nums):
    """与えられた数値のリストが幸運かどうかを返します。ラッキーリストには、7で割り切れる数字が少なくとも1つ含まれています。
    """
    for num in nums:
        if num % 7 == 0:
            return True
        else:
            return False

Try to identify the bug and fix it in the cell below:

バグを特定し、下のセルで修正してみてください。

In [4]:
def has_lucky_number(nums):
    """与えられた数値のリストが幸運かどうかを返します。ラッキーリストには、7で割り切れる数字が少なくとも1つ含まれています。
    """
    for num in nums:
        if num % 7 == 0:
            return True
        
    return False

# Check your answer
q1.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct:</span> 

Remember that `return` causes a function to exit immediately. So our original implementation always ran for just one iteration. We can only return `False` if we've looked at every element of the list (and confirmed that none of them are lucky). Though we can return early if the answer is `True`:

```python
def has_lucky_number(nums):
    for num in nums:
        if num % 7 == 0:
            return True
    # We've exhausted the list without finding a lucky number
    return False
```

Here's a one-line version using a list comprehension with Python's `any` function (you can read about what it does by calling `help(any)`):

```python
def has_lucky_number(nums):
    return any([num % 7 == 0 for num in nums])
```


In [5]:
q1.hint()
q1.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> How many times does the body of the loop run for a list of length n? (If you're not sure, try adding a `print()` call on the line before the `if`.)

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> Remember that `return` causes a function to exit immediately. So our original implementation always ran for just one iteration. We can only return `False` if we've looked at every element of the list (and confirmed that none of them are lucky). Though we can return early if the answer is `True`:

```python
def has_lucky_number(nums):
    for num in nums:
        if num % 7 == 0:
            return True
    # We've exhausted the list without finding a lucky number
    return False
```

Here's a one-line version using a list comprehension with Python's `any` function (you can read about what it does by calling `help(any)`):

```python
def has_lucky_number(nums):
    return any([num % 7 == 0 for num in nums])
```


# 2.
Look at the Python expression below. What do you think we'll get when we run it? When you've made your prediction, uncomment the code and run the cell to see if you were right.

# 2.
次のPython式を見てください。実行すると何が得られると思いますか?予測が完了したら、コードのコメントを解除し、セルを実行して予測が正しいかどうかを確認します。

In [7]:
[1, 2, 3, 4] > 2

TypeError: '>' not supported between instances of 'list' and 'int'

R and Python have some libraries (like numpy and pandas) compare each element of the list to 2 (i.e. do an 'element-wise' comparison) and give us a list of booleans like `[False, False, True, True]`. 

Implement a function that reproduces this behaviour, returning a list of booleans corresponding to whether the corresponding element is greater than n.

RとPythonには、numpyやpandasのようなライブラリがあり、リストの各要素を2 (つまり'element-wise' の比較) と比較し、` [False, False, True, True] `のようなブール値のリストを提供してくれます。

この動作を再現する関数を実装し、対応する要素がnより大きいかどうかに対応するブール値のリストを返します。

In [8]:
def elementwise_greater_than(L, thresh):
    """Lと同じ長さのリストを返します。インデックスiの値は、L [i] がthreshより大きい場合にTrueになり、それ以外の場合はFalseになります。
   
    >>> elementwise_greater_than([1, 2, 3, 4], 2)
    [False, False, True, True]
    """
    return [True if thresh < i else False for i in L]
    

# Check your answer
q2.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct:</span> 

Here's one solution:
```python
def elementwise_greater_than(L, thresh):
    res = []
    for ele in L:
        res.append(ele > thresh)
    return res
```

And here's the list comprehension version:
```python
def elementwise_greater_than(L, thresh):
    return [ele > thresh for ele in L]
```


In [9]:
q2.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> Here's one solution:
```python
def elementwise_greater_than(L, thresh):
    res = []
    for ele in L:
        res.append(ele > thresh)
    return res
```

And here's the list comprehension version:
```python
def elementwise_greater_than(L, thresh):
    return [ele > thresh for ele in L]
```


# 3.

Complete the body of the function below according to its docstring.

# 3.

docstringに従って、以下の関数の本体を完成させます。

In [15]:
def menu_is_boring(meals):
    """一定期間に提供された食事のリストを指定し、同じ食事が2日連続で提供された場合はTrueを返し、それ以外の場合はFalseを返します。
    """
    if len(meals) < 1:
        return False
        
    kari = meals[0]
    for i in range(1,len(meals)):
        if kari == meals[i]:
            return True
        kari = meals[i]
    return False
        
        

# Check your answer
q3.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct:</span> 



```python
def menu_is_boring(meals):
    # Iterate over all indices of the list, except the last one
    for i in range(len(meals)-1):
        if meals[i] == meals[i+1]:
            return True
    return False
```

The key to our solution is the call to `range`. `range(len(meals))` would give us all the indices of `meals`. If we had used that range, the last iteration of the loop would be comparing the last element to the element after it, which is... `IndexError`! `range(len(meals)-1)` gives us all the indices except the index of the last element.

But don't we need to check if `meals` is empty? Turns out that `range(0) == range(-1)` - they're both empty. So if `meals` has length 0 or 1, we just won't do any iterations of our for loop.


In [16]:
q3.hint()
q3.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> This is a case where it may be preferable to iterate over the *indices* of the list (using a call to `range()`) rather than iterating over the elements of the list itself. When indexing into the list, be mindful that you're not "falling off the end" (i.e. using an index that doesn't exist).

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 

```python
def menu_is_boring(meals):
    # Iterate over all indices of the list, except the last one
    for i in range(len(meals)-1):
        if meals[i] == meals[i+1]:
            return True
    return False
```

The key to our solution is the call to `range`. `range(len(meals))` would give us all the indices of `meals`. If we had used that range, the last iteration of the loop would be comparing the last element to the element after it, which is... `IndexError`! `range(len(meals)-1)` gives us all the indices except the index of the last element.

But don't we need to check if `meals` is empty? Turns out that `range(0) == range(-1)` - they're both empty. So if `meals` has length 0 or 1, we just won't do any iterations of our for loop.


# 4. <span title="A bit spicy" style="color: darkgreen ">🌶️</span>

Next to the Blackjack table, the Python Challenge Casino has a slot machine. You can get a result from the slot machine by calling `play_slot_machine()`. The number it returns is your winnings in dollars. Usually it returns 0.  But sometimes you'll get lucky and get a big payday. Try running it below:

# 4. <span title="A bit spicy" style="color: darkgreen ">🌶️</span>

ブラックジャックのテーブルの隣には、パイソンチャレンジカジノのスロットマシンがあります。`play_slot_machine () `を呼び出すことで、スロットマシンから結果を取得できます。返される数字はドルでの賞金です。通常は0を返します。でも、運が良くて大金を手にすることもあります。以下で実行してみてください。

In [19]:
play_slot_machine()

1.5

By the way, did we mention that each play costs $1? Don't worry, we'll send you the bill later.

On average, how much money can you expect to gain (or lose) every time you play the machine?  The casino keeps it a secret, but you can estimate the average value of each pull using a technique called the **Monte Carlo method**. To estimate the average outcome, we simulate the scenario many times, and return the average result.

Complete the following function to calculate the average value per play of the slot machine.

ところで、1回のプレイに1ドルかかると言いましたか?後ほど請求書をお送りしますのでご安心ください。

平均して、ゲーム機をプレイするたびにどれくらいのお金を得る (または失う) ことができますか?カジノはそれを秘密にしていますが、**モンテカルロ法と呼ばれる手法を使用して、各プルの平均値を推定できます**.平均結果を推定するには、シナリオを何度もシミュレートし、平均結果を返します。

スロットマシンの1プレイあたりの平均値を計算するには、次の関数を完成させます。

In [24]:
def estimate_average_slot_payout(n_runs):
    """スロットマシンをn回実行し、1回あたりの平均純利益を返します。
    呼び出し例:戻り値は非決定的であることに注意してください。
    >>> estimate_average_slot_payout(1)
    -1
    >>> estimate_average_slot_payout(1)
    0.5
    """
    payouts = [play_slot_machine()-1 for i in range(n_runs)]
    return sum(payouts) / n_runs

In [29]:
estimate_average_slot_payout(10000000)

0.0259772

When you think you know the expected value per spin, run the code cell below to view the solution and get credit for answering the question.

スピンあたりの期待値がわかったら、次のコードセルを実行してソリューションを表示し、質問に回答するとクレジットを取得します。

In [30]:
# 答えをチェック(クレジットを受け取るには、このコードセルを実行してください。)
q4.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
    
The exact expected value of one pull of the slot machine is 0.025 - i.e. a little more than 2 cents.  See?  Not every game in the Python Challenge Casino is rigged against the player!

In order to get this answer, you'll need to implement the `estimate_average_slot_payout(n_runs)` function to simulate pulling the slot machine `n_runs` times.  It should return the payout averaged over those `n_runs`.

Then, once the function is defined, in order to estimate the average slot payout, we need only call the function.

Because of the high variance of the outcome (there are some very rare high payout results that significantly affect the average) you might need to run your function with a very high value of `n_runs` to get a stable answer close to the true expectation.  For instance, you might use a value for `n_runs` of 1000000.

Here's an example for how the function could look:
```python
def estimate_average_slot_payout(n_runs):
    # Play slot machine n_runs times, calculate payout of each
    payouts = [play_slot_machine()-1 for i in range(n_runs)]
    # Calculate the average value
    avg_payout = sum(payouts) / n_runs
    return avg_payout
    
estimate_average_slot_payout(10000000)

```

This should return an answer close to 0.025!
            


解決方法:

スロットマシンの1回の引き当たりの正確な期待値は0.025、つまり2セントを少し上回る程度です。ほらね?Pythonチャレンジカジノのすべてのゲームがプレイヤーに対して仕掛けられているわけではありません。

この答えを得るには、estimate_average_slot_payout (n_runs) 関数を実装して、スロットマシンをn_runs回引き出すことをシミュレートする必要があります。これらのn_回の平均配当を返す必要があります。

関数が定義されたら、スロットの平均ペイアウトを推定するために、関数を呼び出すだけで済みます。

結果の分散が大きいため (平均に大きく影響する非常にまれな高配当の結果もあります)、真の期待値に近い安定した結果を得るには、n_runsの値を非常に大きくして関数を実行する必要がある場合があります。たとえば、n_runsの値として100万を使用できます。

関数がどのようになるかの例を示します。

```Python
def estimate_average_slot_payout (n_runs) :
    # スロットマシンをn回プレイし、それぞれのペイアウトを計算
    payouts= [play_slot_machine () -1 for i in range (n_runs) ]
    # 平均値を計算する
    avg_payout = sum(payouts) / n_runs
    return avg_payout
```

estimate_average_slot_payout (1000万)
これにより、0.025に近い答えが返されるはずです。

# Keep Going

Many programmers report that dictionaries are their favorite data structure. You'll get to **[learn about them](https://www.kaggle.com/colinmorris/strings-and-dictionaries)** (as well as strings) in the next lesson.

# 続けて

多くのプログラマは、辞書がお気に入りのデータ構造であると報告しています。次のレッスンでは、** [learn about them](https://www.kaggle.com/colinmorris/strings-and-dictionaries)** (および文字列) について説明します。

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/python/discussion) to chat with other learners.*

---





*ご質問やご意見がありますか?[Course Discussion Forum] (https://www.kaggle.com/learn/python/discussion) にアクセスして、他の学習者とチャットしてください。*